SNOW-508732 Update Sink Service EOS

sfc-gh-anwang opened the pull request:
When EOS is enabled: 
- Call configure client and get status APIs when initializing ServiceContext
- Discard the record if the record offset is smaller or equal to offset persisted in snowflake
- Insert file api also contains client info (clientSequencer and offsetToken)

sfc-gh-japatel commented:
Can we please add description in this PR?

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -432,6 +434,10 @@ protected static String getNameIndex(String topic, int partition) {
 
+    // exactly once semantics
+    private final AtomicLong clientSequencer = new AtomicLong(-1);
+    private final AtomicLong serverSideOffset = new AtomicLong(-1);
```
may be something like `offsetPersistedInSnowflake`

Can you also add a comment saying `this is source of truth` Also this value doesnt necessarily mean it is available to query but just a notification that snowflake has received this offset and has queued it?

sfc-gh-japatel replied:
For verifying whether that offset is ingested, we still have to rely on insertReport api

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -493,6 +501,23 @@ private void init(long recordOffset) {
       recover(pipeCreation);
 
+      // when exactly_once is enabled,fetch clientSequencer and offsetPersistedInSnowflake
+      if (ingestionDeliveryGuarantee
```
lets extract this out to a new method so that this init method is not noisy. 
Also since it calls only during start, we should log the offsetToken and clientSequencer.

sfc-gh-anwang replied:
Sounds good

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -813,7 +842,19 @@ private long getOffset() {
+        ingestionService.ingestFilesWithClientInfo(fileNamesCopy, this.clientSequencer.get());
+        String offsetToken = ingestionService.getClientStatus().getOffsetToken();
+        // Update server side offset
+        if (offsetToken == null) {
+          this.offsetPersistedInSnowflake.set(-1);
```
Is this even possible? Should we throw an exception if offsetToken we receive from SF is null?

sfc-gh-anwang replied:
when we create new pipe, the offset token is null when we call get pipe status

sfc-gh-anwang replied:
we can not put null offset token in insert call though

sfc-gh-japatel replied:
> when we create new pipe, the offset token is null when we call get pipe status

This will not happen in this call method right? because we are calling `ingestFilesWithClientInfo`

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -813,7 +842,19 @@ private long getOffset() {
+        // Update server side offset
+        if (offsetToken == null) {
+          this.offsetPersistedInSnowflake.set(-1);
+        } else {
+          this.offsetPersistedInSnowflake.set(Long.parseLong(offsetToken));
```
Just thinking out loud here. Why do we want to set this value after returning from another rest call for getClientStatus? 

Cant we rely on `insertFilesWithClientInfo`'s successful call? Basically if it is a success, the `this.offsetPersistedInSnowflake.set(lastOffsetFromFrom_fileNamesCopy)`

I am okay with one more though for completeness.

sfc-gh-anwang replied:
if we only need the offsetPersistedInSnowflake during rebalancing, this call is kind of redundant already. i am also ok with directly setting the offset to the lastOffsetFromFilename.

sfc-gh-japatel replied:
We can eventually get away with it if we have enough data about its completeness. Can you put a log here?

sfc-gh-japatel replied:
Also, lets verify we have a logger in `insertFilesWithClientInfo's` on what we are passing in as offsetToken.. that way we can decide if this is a redundant call

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -651,10 +676,14 @@ private void insert(final SinkRecord record) {
+      // else discard the record if the record offset is smaller or equal to server side offset
+      if ((ingestionDeliveryGuarantee
+                  == SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE
+              || record.kafkaOffset() > this.offsetPersistedInSnowflake.get())
+          && record.kafkaOffset() > processedOffset.get()) {
```
in case of exactly_once, can we set `processedOffset` = `offsetPersistedInSnowflake`?
Then we can avoid this complicated if logic!

sfc-gh-anwang replied:
proceedOffset can sometime be more update to date than the offsetPersistedInSnowflake, so this might not actually work

sfc-gh-japatel replied:
Thank you, the idea here is to make the new code as little as possible to invoke when default/exactly once is set.

sfc-gh-japatel commented on src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java:
```diff
@@ -1147,4 +1148,229 @@ public void testCleanerRecoverListCount() throws Exception {
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    service2.insert(record2);
+    // The record should not be inserted in the buffer
+    Assert.assertTrue(
+        ((SnowflakeSinkServiceV1) service2)
+            .isPartitionBufferEmpty(SnowflakeSinkServiceV1.getNameIndex(topic, partition)));
```
nice!

sfc-gh-japatel commented on src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java:
```diff
@@ -1147,4 +1148,229 @@ public void testCleanerRecoverListCount() throws Exception {
+                SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE)
+            .addTask(table, topic, partition)
+            .build();
+    offset = 1;
+    // Create sink record with same offset
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    // Should update the client sequencer
+    service2.insert(record2);
+    // call snowpipe ingest api
+    service.callAllGetOffset();
+
+    service.closeAll();
+    service2.closeAll();
```
but we are changing the offset here to 1? So it should ingest it?

sfc-gh-anwang replied:
Sorry, the comment wasn't very clear, the old service calls get offset, so the api call should throw exception due to outdated client sequencer.

sfc-gh-japatel commented on src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java:
```diff
@@ -1147,4 +1148,229 @@ public void testCleanerRecoverListCount() throws Exception {
+    // Should update the client sequencer
+    service2.insert(record2);
+    // call snowpipe ingest api
+    service.callAllGetOffset();
```
so this will throw an exception? Can you catch the exception and assert the exception type? I think we discussed that it will throw IngestException so may be we can just catch it? 

And then may be we can call  `service2.callAllGetOffset();` to see if offset 1 is getting ingested? 

In an ideal world we would expect the offset 0 to be ingested by `service2` (The newly woken up instance after rebalance) since precommit failed in `service0` and offset 0 was never ingested in SF.

sfc-gh-japatel replied:
I ran this test. we should catch exception and assert. 
I see all retries with 
```Message: Pipe has older version of clientSequencer. Please try again with latest clientSequencer.,
Data: null
}```
Which seems corrcet and behavior also seemed to be correct.

i.e on this exception technically there is another node with higher sequencer trying to ingest the data. so this instance of connector should stop working. i.e stop abruptly.

sfc-gh-japatel reviewed (commented):
no major changes but will wait for one more pass. This is great stuff Andy. Good to see all tests being added.

sfc-gh-japatel commented:
Also, we might have to add google guava dependency 31.0jre since we removed that dependency from snowflake-java sdk . 

We can check that too in pom.xml and confluent_pom.xml

```
<dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>31.0.1-jre</version>
        </dependency>
```

sfc-gh-anwang commented:
> Also, we might have to add google guava dependency 31.0jre since we removed that dependency from snowflake-java sdk .
> 
> We can check that too in pom.xml and confluent_pom.xml
> 
> ```
> <dependency>
>             <groupId>com.google.guava</groupId>
>             <artifactId>guava</artifactId>
>             <version>31.0.1-jre</version>
>         </dependency>
> ```

I added this dependency on my local branch, I'll check that in then.

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -526,6 +515,26 @@ private void init(long recordOffset) {
       }
     }
 
+    private void initClientInfo() {
+      ConfigureClientResponse configureClientResponse = ingestionService.configureClient();
```
A short description here would be very helpful. I know we usually dont put Javadocs on private method but it's just easier to follow the code. 

Also lets rename this to `initClientInfoForExactlyOnceDelivery`

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:
```diff
@@ -526,6 +515,26 @@ private void init(long recordOffset) {
+        } else {
+          this.offsetPersistedInSnowflake.set(Long.parseLong(offsetToken));
+        }
+        logInfo(
+            "Initialized client info, clientSequencer: {}, offsetToken: {}.",
+            this.clientSequencer.get(),
+            this.offsetPersistedInSnowflake.get());
```
Awesome. can we also print pipeName from servicecontext?

sfc-gh-japatel commented on src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:
```diff
@@ -127,8 +127,8 @@ static void setFieldToDefaultValues(Map<String, String> config, String field, Lo
-   * @param key name of the key to be retrived
-   * @return proprity value or null
+   * @param key name of the key to be retrieved
+   * @return property value or null
```
thank you!

sfc-gh-japatel reviewed (approved):
lgtm! Way to go! 

@sfc-gh-anwang and I have performed End to End test locally and we will push this once we release 5.44 server side and we release Snowflake-ingest-java next version. 

@sfc-gh-anwang can you file a JIRA to add a new template to `/test/rest_request_template/` directory with `exactly_once`?
